src/librustc/mir/tcx.rs

```diff
@@ -119,6 +119,39 @@ impl<'tcx> Place<'tcx> {
                 proj.base.ty(local_decls, tcx).projection_ty(tcx, &proj.elem),
         }
     }
+
+    /// If this is a field projection, and the field is being projected from a closure type,
+    /// then returns the index of the field being projected. Note that this closure will always
+    /// be `self` in the current MIR, because that is the only time we directly access the fields
+    /// of a closure type.
+    pub fn is_upvar_field_projection<'cx, 'gcx>(&self, mir: &'cx Mir<'tcx>,
+                                                tcx: &TyCtxt<'cx, 'gcx, 'tcx>,
+                                                recurse: bool) -> Option<Field> {
+        match *self {
+            Place::Projection(ref proj) => {
+                match proj.elem {
+                    ProjectionElem::Field(field, _ty) => {
+                        let base_ty = proj.base.ty(mir, *tcx).to_ty(*tcx);
+
+                        if  base_ty.is_closure() || base_ty.is_generator() {
+                            Some(field)
+                        } else {
+                            None
+                        }
+                    },
+                    ProjectionElem::Deref => {
+                        if recurse {
```

Comment:
I wasn't sure about adding this parameter - without it, a handful of unrelated tests start throwing duplicate errors with shorter spans.

Reply:
I've just checked and it seems like only doing a single deref still exhibits this behaviour.

Reply:
Pushed a fix for this.